### ISSUE TITLE
KenCarp4 performance with various Jacobians on first run

### ISSUE THREAD

**Issue by isaacsas** [SciML/OrdinaryDiffEq.jl#722]
Feel free to close if this is better discussed elsewhere. I just wanted to report what I'm seeing on KenCarp4 for the BCR network with a finite difference Jacobian, an explicit dense Jacobian, and an explicit sparse Jacobian (using `lu` as the linear solver). Here `f1` and `f2` are pre-solve calls to the ODE rhs and `J1` and `J2` are pre-solve calls to the Jacobian evaluation function. What I'm trying to understand is why the first solve with either the dense or sparse Jacobian seems to take much longer than the second despite pre-evaluating the Jacobian function (in contrast to using the finite difference approximation). For the dense case I did also try `@noinline` in some of my testing and saw no difference (though I'm not sure if I checked this specific simulation/parameters). I can check it again if that is a concern.

Note, these were each run in a fresh REPL.

Dense Jacobian with default linear solver:
```
                              Time                   Allocations      
                      ──────────────────────   ───────────────────────
   Tot / % measured:        644s / 100%            17.2GiB / 98.9%    

 Section      ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────
 J1                1     488s  76.1%    488s   7.75GiB  45.5%  7.75GiB
 KenCarp4-1        1    65.4s  10.2%   65.4s   4.71GiB  27.6%  4.71GiB
 f1                1    41.6s  6.49%   41.6s   3.42GiB  20.0%  3.42GiB
 KenCarp4-2        1    25.0s  3.89%   25.0s    111MiB  0.63%   111MiB
 J2                1    886μs  0.00%   886μs     32.0B  0.00%    32.0B
 f2                1   82.4μs  0.00%  82.4μs     32.0B  0.00%    32.0B
 ─────────────────────────────────────────────────────────────────────
```
No Jacobian solve (finite difference):
```
                              Time                   Allocations      
                      ──────────────────────   ───────────────────────
   Tot / % measured:        124s / 98.5%           5.76GiB / 96.9%    

 Section      ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────
 f1                1    40.9s  33.5%   40.9s   3.42GiB  61.2%  3.42GiB
 KenCarp4-1        1    38.6s  31.6%   38.6s   1.35GiB  24.1%  1.35GiB
 KenCarp4-2        1    32.4s  26.5%   32.4s    111MiB  1.94%   111MiB
 f2                1   80.8μs  0.00%  80.8μs     32.0B  0.00%    32.0B
 ─────────────────────────────────────────────────────────────────────
```
Explicit sparse LU with `lu` as the linear solver:
```
                             Time                   Allocations      
                      ──────────────────────   ───────────────────────
   Tot / % measured:        693s / 100%            29.0GiB / 99.4%    

 Section      ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────
 J1                1     478s  69.2%    478s   7.00GiB  24.3%  7.00GiB
 KenCarp4-1        1    95.3s  13.8%   95.3s   11.0GiB  38.2%  11.0GiB
 KenCarp4-2        1    57.1s  8.27%   57.1s   6.37GiB  22.1%  6.37GiB
 f1                1    40.8s  5.90%   40.8s   3.42GiB  11.9%  3.42GiB
 J2                1    158μs  0.00%   158μs     32.0B  0.00%    32.0B
 f2                1   81.0μs  0.00%  81.0μs     32.0B  0.00%    32.0B
 ─────────────────────────────────────────────────────────────────────
```
The basic solve command I used was 
```julia
@timeit to "KenCarp4-2" begin bsol = solve(boprob,KenCarp4(autodiff=false,linsolve=LinSolveFactorize(lu)), saveat=1., abstol=1e-8, reltol=1e-8); end; show(to)
```
for the sparse case and 
```julia
@timeit to "KenCarp4-1" begin sol = solve(boprob,KenCarp4(autodiff=false), abstol=1e-8, reltol=1e-8, saveat=1.); end; show(to)
```
for the dense/FD case.

**Comment by isaacsas** [comment]
If I have time this weekend I'll try profiling a shorter run to see if there is anything obvious going on.

**Comment by YingboMa** [comment]
Could you post your source code in the GitHub Gist?

**Comment by isaacsas** [comment]
I’ll put a script up for you later today.

**Comment by isaacsas** [comment]
Messaged you on Slack, but I'll put a copy here for posterity. You need the latest DiffEqBio master and to run from the examples folder of ReactionNetworkImporters (or modify the path in the script):

https://gist.github.com/isaacsas/7382521eb9c1b9b36a29877e5d2fdaf0

**Comment by isaacsas** [comment]
@ChrisRackauckas Just recording what DEStats gives on the iterative FD J*v: (note, I've played around with different gmres `tol` and `maxiter` -- I'm not seeing other choices give a noticeable speedup).

With `jac_prototype=JacVecOperator{Float64}(f,u0,p,0.; autodiff=false)` and
```julia
@timeit to "KenCarp4-1" begin bsol = solve(boprob,KenCarp4(autodiff=false, linsolve=LinSolveGMRES(tol=1e-11,restart=5,maxiter=5)), abstol=1e-8, reltol=1e-8, saveat=1.); end; show(to)
```
I'm getting
```
┌ Warning: Interrupted. Larger maxiters is needed.
└ @ DiffEqBase ~/.julia/packages/DiffEqBase/EFqMn/src/integrator_interface.jl:150
 ─────────────────────────────────────────────────────────────────────
                              Time                   Allocations      
                      ──────────────────────   ───────────────────────
   Tot / % measured:       1393s / 96.8%            221GiB / 100%     

 Section      ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────
 KenCarp4-1        1    1348s   100%   1348s    221GiB  100%    221GiB
 ─────────────────────────────────────────────────────────────────────

DiffEqBase.DEStats
Number of function 1 evaluations: 3254818
Number of function 2 evaluations: 0
Number of W matrix evaluations: 999996
Number of linear solves:        3488892
Number of Jacobians created:        0
Number of nonlinear solver iterations:        3254815
Number of nonlinear solver convergence failures:        765923
Number of rootfind condition calls:        0
Number of accepted steps:       234077
Number of rejected steps:       0
```
When I plot the solution up to the (premature) termination time it looks (qualitatively) fine.

For reference, KenCarp4 with dense LU finite difference (no Jacobian given),
```julia
@timeit to "KenCarp4-1" begin bsol = solve(boprob,KenCarp4(autodiff=false), abstol=1e-8, reltol=1e-8, saveat=1.); end; show(to)
```
is < 40s and gives
```
DiffEqBase.DEStats
Number of function 1 evaluations: 221600
Number of function 2 evaluations: 0
Number of W matrix evaluations: 431
Number of linear solves:        34680
Number of Jacobians created:        168
Number of nonlinear solver iterations:        32933
Number of nonlinear solver convergence failures:        154
Number of rootfind condition calls:        0
Number of accepted steps:       1747
Number of rejected steps:       0
```

**Comment by isaacsas** [comment]
Perhaps this is related to not giving gmres a starting guess?

**Comment by ChrisRackauckas** [comment]
If it is, then we need to more explicitly handle a `krylov=true` thing in NLNewton so we can hold that kind of information.

**Comment by ChrisRackauckas** [comment]
@isaacsas what about KenCarp4 with GMRES and not using the jac_prototype? Isolate if that's the issue.

**Comment by isaacsas** [comment]
I don’t the issue is the starting guess. I was reading a bit about Newton-Krylov methods from the review by Knoll and Keyes last night  — they suggest exactly what you did, using zero as the starting guess for the linear solver within the Newton method.

**Comment by isaacsas** [comment]
Here's a link to the review: https://www.cs.odu.edu/~keyes/papers/jfnk.pdf

If I don't set a jac_prototype or Jacobian function does that mean it will use finite differences to approximate the full Jacobian with gmres?

**Comment by ChrisRackauckas** [comment]
>If I don't set a jac_prototype or Jacobian function does that mean it will use finite differences to approximate the full Jacobian with gmres?

yup.

**Comment by isaacsas** [comment]
It's running, still clearly much slower than the dense LU with FD case. I'll post if it ends up finishing noticeably faster than with a prototype but I'm doubtful.  This is with maxiter=restart=5 in gmres, which I think is consistent with Sundials. I can play around with that more if later. I also switched back to the epsilon choice that is currently in OrdinaryDiffEq (I played a bit with a weighted epsilon like in Sundials).

**Comment by isaacsas** [comment]
Yeah, it's still super slow so that seems to indicate the problem isn't the J*v code.

**Comment by ChrisRackauckas** [comment]
Where are we at on this? Some investigations noted in https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/724 shows there's no extra tricks in Sundials here, no initial condition or Krylov subspace keeping (though that would be interesting to investigate). 
So can I get a summary of where we are? It seems that:

1. Matrix-free finite differencing fails
2. Standard finite differencing works (but is slow, of course)
3. AD can't run on this case.

Those facts would point to our finite differencing being numerically unstable, with our GMRES being fine?

**Comment by isaacsas** [comment]
I've been meaning to play with this more, but had a combination of things keep me from having any time to work on it the past month. If I'm remembering correctly Sundials does two things we currently don't have; inexact Newton-Krylov (i.e. updating the matrix free-operator each step of the Newton solver), and solving a rescaled linear system with GMRES (essentially rescaling by the weights used in Sundials' weighted error norm -- I think these weights would also need to be updated at each step of the Newton solver as they are solution dependent). Sundials may also use different error measures for stopping the Newton iterations and the Krylov iterations, with a fixed relationship between their associated tolerances (for example, I think they use the absolute value of the residual in GMRES, but IterativeSolvers may use a relative residual).

I tried modifying the current nlsolver to update the W matrix each step as we had discussed. For the BCR system using matrix-free finite differences with GMRES and no rescaling/preconditioning I think the solver terminated early with stability warnings (and the performance was pretty poor). I wasn't able to get it to ever complete a solve. It could be I didn't implement the updates correctly, or perhaps the linear system rescaling that Sundials uses is necessary. I didn't get a chance to try to implement such a rescaling and see how it changed convergence. In contrast, Sundials with inexact finite-difference Newton-Krylov and no explicitly specified preconditioner works and gives an accurate answer (though a bit slower than when using Sundials with LU).

For the current nlsolver (i.e. modified Newton method) standard finite differencing with dense LU works and does slightly beat the CVODE equivalent. Everything else I tried was either exceptionally slow, gave max iteration warnings, or gave stability warnings.
